src/mongocxx/test/spec/unified_tests/operations.cpp

```diff
@@ -1068,54 +1096,54 @@ document::value update_one(collection& coll, client_session* session, document::
         options.upsert(arguments["upsert"].get_bool().value);
     }
 
-    std::int32_t matched_count = 0;
-    bsoncxx::stdx::optional<std::int32_t> modified_count;
-    bsoncxx::stdx::optional<types::bson_value::view> upserted_id{};
+    stdx::optional<result::update> update_one_result;
 
-    if (session) {
-        auto update_one_result = coll.update_one(*session, filter, update, options);
-        if (update_one_result) {
-            matched_count = update_one_result->matched_count();
-            modified_count = update_one_result->modified_count();
+    {
+        const auto update = arguments["update"];
+        const auto type = update.type();
 
-            if (auto upserted_element = update_one_result->upserted_id()) {
-                upserted_id = upserted_element->get_value();
+        switch (type) {
+            case bsoncxx::type::k_document: {
+                const auto doc = update.get_document().value;
+                update_one_result = session ? coll.update_one(*session, filter, doc, options)
+                                            : coll.update_one(filter, doc, options);
+                break;
             }
-        }
-    } else {
-        auto update_one_result = coll.update_one(filter, update, options);
-        if (update_one_result) {
-            matched_count = update_one_result->matched_count();
-            modified_count = update_one_result->modified_count();
-
-            if (auto upserted_element = update_one_result->upserted_id()) {
-                upserted_id = upserted_element->get_value();
+            case bsoncxx::type::k_array: {
+                pipeline pipe;
+                pipe.append_stages(update.get_array().value);
+                update_one_result = session ? coll.update_one(*session, filter, pipe, options)
+                                            : coll.update_one(filter, pipe, options);
+                break;
             }
+            default:
+                throw std::logic_error("unexpected type '" + to_string(type) +
+                                       "' for field 'update'");
         }
     }
 
     auto result = builder::basic::document{};
-    result.append(builder::basic::kvp(
-        "result",
-        [matched_count, modified_count, upserted_id](builder::basic::sub_document subdoc) {
-            subdoc.append(builder::basic::kvp("matchedCount", matched_count));
+    result.append(builder::basic::kvp("result", [&](builder::basic::sub_document subdoc) {
+        if (!update_one_result) {
+            subdoc.append(builder::basic::kvp("matchedCount", 0));
```

Comment:
Suggest removing the `matchedCount`.

```suggestion
```

`update_one_result` is [only expected to be unset if an unacknowledged write concern is used](https://github.com/mongodb/mongo-cxx-driver/blob/09d9d7e736b86724c21e4e41e57c1112ebf5656b/src/mongocxx/collection.hpp#L1611-L1613).

Returning `"matchedCount": 0` suggests that the server replied. An unacknowledged write concern does not have a server reply.

The [updateOne-hint-unacknowledged.yml](https://github.com/mongodb/specifications/blob/ba44473e2488290e13acabbea0f32fa8c2bc4b7a/source/crud/tests/unified/updateOne-hint-unacknowledged.yml#L69) test has an assertion on the result of an unacknowledged update. It only checks for the optional `acknowledged` field.

Reply:
Will apply suggestion.

Note: this line was included to preserve this prior behavior where `matchedCount` was unconditionally included in the result:

```cpp
std::int32_t matched_count = 0;
// ...

if (session) {
    auto update_one_result = coll.update_one(*session, filter, update, options);
    if (update_one_result) {
        matched_count = update_one_result->matched_count();
        // ...
    }
} else { /* ... */ }

auto result = builder::basic::document{};
result.append(builder::basic::kvp(
    "result",
    [/* ... */](builder::basic::sub_document subdoc) {
        subdoc.append(builder::basic::kvp("matchedCount", matched_count));
        // ...
    }));
```